enhancements/installer/connected-assisted-installer.md

```diff
@@ -0,0 +1,323 @@
+---
+title: connected-assisted-installer
+authors:
+  - "@avishayt"
+  - "@hardys"
+  - "@dhellmann"
+reviewers:
+  - "@beekhof"
+  - "@deads2k"
+  - "@hexfusion"
+  - "@mhrivnak"
+approvers:
+  - "@crawford"
+  - "@abhinavdahiya"
+  - "@eparis"
+creation-date: 2020-06-09
+last-updated: 2020-06-10
+status: implementable
+see-also:
+  - "/enhancements/baremetal/minimise-baremetal-footprint.md"
+---
+
+# Assisted Installer for Connected Environments
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement describes changes in and around the installer to
+assist with deployment on user-provisioned infrastructure. The use
+cases are primarily relevant for bare metal, but in the future may be
+applicable to cloud users who are running an installer UI directly
+instead of using a front-end such as `cloud.redhat.com` or the UI
+provided by their cloud vendor.
+
+## Motivation
+
+The target user is someone wanting to deploy OpenShift, especially on
+bare metal, with as little up-front infrastructure dependencies as
+possible. This person has access to servers and wants to run workloads
+quickly. They do not necessarily have the administrative privileges to
+create private VLANs, configure DHCP/PXE servers, or manage other
+aspects of the infrastructure surrounding the hardware where the
+cluster will run. If they do have the required privileges, they may
+not want to delegate them to the OpenShift installer for an IPI
+installation, preferring instead to use their existing tools and
+processes for some or all of that configuration. They are willing to
+accept that the cluster they build may not have all of the same
+features as an IPI cluster on day 1, but that by taking additional
+steps later they will be able to add more automatic infrastructure
+management and make it fully production-ready.
+
+### Goals
+
+- Make initial deployment of usable and supportable clusters simpler.
+- Move more infrastructure configuration from day 1 to day 2.
+- Support connected on-premise deployments.
+- Support existing IPI features, especially for day 2 cluster
```

Comment:
nit: provisioner shouldn't matter for features, especially day-2 features.  This should be:

> * Support cluster-managed infrastructure features, especially on day 2.

Or something like that...

Reply:
I'm not aware that it is currently possible to install a UPI bare metal cluster and then on day 2 use IPI to add workers. That's the goal here.

Reply:
There are features that you get today on bare metal only if you used IPI and do not get with UPI, particularly power management via BMC. Given the direction we're going, it probably makes sense for us to stop referring to that as "existing IPI features" and get in the habit of using a more direct descriptor.

Reply:
Are we really just disagreeing about the wording? Or is the description unclear?

Reply:
I had to look up [baseboard management controller][1], so take this with a grain of "Trevor has no idea what he's talking about" salt ;).  But it seems to me that "I have an existing metal cluster and want to give its machine API controller the ability and authority to manage Machines as a day 2 operator" would be a different enhancement from (and possibly a prerequisite for) this install-time enhancement.  Couldn't you set up that day-2 management without needing to call out to `cloud.redhat.com`?  You'd already have a local cluster where you could host whatever the incoming machines would need for their first boot...

[1]: https://en.wikipedia.org/wiki/Intelligent_Platform_Management_Interface#Baseboard_management_controller

Reply:
You're right, the two things are somewhat related.

The existing bare metal IPI workflow requires management controller credentials for each host, so the installer can automate the deployment. Under that flow the installer has all of the information that would be required to turn the machine API on, so it does not make much sense to assume the machine API is disabled and we expect to always leave it enabled.

It also should be possible to use the UPI process to install a cluster and then add the machine API integration later. There's a separate set of work planned for that, regardless of the progress on this enhancement. For a UPI process, the user is going to have to do more work to set up the machine API, including BareMetalHost CRs with specific settings and connecting them to Machine CRs.

The important thing we're referring to here is that we can save some of the information available during the new automated installation process so that the user has less work to do to enable the machine API later. It seemed like adding the details of how that would work would be going too deep into planned implementation, but if we need more detail in this document about how that would work we can add it.

Reply:
The current interface which decides if cluster-managed-baremetal pieces are enabled is the platform, so when you do an IPI deployment using `platform: baremetal` you get those features.

However the current UPI docs use `platform: none` and there's no currently supported way to enable the managed-baremetal pieces after initial deployment.  There has been some work looking at how we might change that ref https://github.com/openshift/machine-api-operator/pull/560

However I don't think making day-2 enablement of managed baremetal is a prerequisite for this assisted-installer workflow, because AFAIK the plan is to use the `platform: baremetal` even in the case where you choose the "boot it yourself" path - this also gives some other features like cluster-hosted loadbalancing and DNS, so it makes sense to reuse the same platform if possible.  /cc @avishat to keep me honest on plans here :)

Reply:
Indeed, we use installer platform: baremetal to generate the ignitions